### ISSUE TITLE
Is there a track-independent way to get the exercise version, currently implemented in a track?

### ISSUE THREAD

**Issue by ZapAnton** [exercism/exercism#4466]
I was planning to make a small utility to get all outdated or unimplemented exercises in the various language tracks.

The way to check if the exercise is outdated would be to get the local version of the exercise and compare it with the `problem-specifications` version.

Unfortunately I could not find the appropriate field in the `config.json` and for now I have to look at how different tracks implements their versioning (e.g. saving exercise version in `package.json` or `Cargo.toml`).

Is there a more universal way to get the local version of the exercise?

**Comment by petertseng** [comment]
According to https://github.com/exercism/exercism/issues/4131#issuecomment-410118808 

> it would not surprise me if other tracks now or in the future also had project-specific canonical version locations. Adding versioning information elsewhere would require either keeping the two version strings synchronized, or depreciating the track-specific location in favor of the exercism-specific one.

Thus, there is no universal location. The current way to deal with this problem is to define the track-specific way to get the version. As you know, the rest of the logic stays the same, regardless of the way of getting the track-specific version

* https://github.com/petertseng/exercism-problem-specifications/blob/up-to-date/up-to-date/ceylon.rb
* https://github.com/petertseng/exercism-problem-specifications/blob/up-to-date/up-to-date/go.rb
* https://github.com/petertseng/exercism-problem-specifications/blob/up-to-date/up-to-date/haskell.rb
* https://github.com/petertseng/exercism-problem-specifications/blob/up-to-date/up-to-date/rust.rb

**Comment by ZapAnton** [comment]
The hard part is that not every track keeps exercise versions. 

Quickly looking through the tracks and this seems to be the case for .NET languages, JVM-based languages (except Ceylon), C, C++ and Crystal to name a few

**Comment by petertseng** [comment]
Indeed; it seems that such tracks would not be able to take advantage.

**Comment by rpottsoh** [comment]
In the Delphi track I am in the habit of including the `CanonicalVersion` as a const near the top of test suite unit.  This was mainly for my benefit to allow me to see quickly if my version was behind the canonical data for the exercise.

**Comment by kytrinyx** [comment]
> see quickly if my version was behind the canonical data for the exercise.

Out of curiosity, in which context are you using that? On the exercism.io website? Or in the course of track maintenance?

**Comment by ZapAnton** [comment]
Track maintenance. Potentially (if the track maintainers will agree) the utility will be added to the track repo, so that new contributors could use it to find new exercises to implement, or outdated exercises to update

**Comment by rpottsoh** [comment]
@kytrinyx good question.  It is for the latter.  I started doing this during V1.

**Comment by kytrinyx** [comment]
@rpottsoh In that case I'd rather not have it as part of the code, but rather as part of the metadata, probably ideally as part of the `.meta/` directory, which doesn't get transmitted to users.

If we need the data on the site then I'd want to put it in `config.json` as part of the exercise objects in the `exercises` array, but I don't think we do.

**Comment by rpottsoh** [comment]
Just so I am clear, there is no problem with what I am doing, for the reason I'm doing it?  I have no qualms with the data also appearing in other places for the sake of the site.

However, my concerns of the version not appearing, in some form, in the test suite itself is that I think it would become quite easy for versionitus to creep in as the version "stamp" for the test suite would no longer be on the test suite itself.  The maintainers now have to remember that their workflow would now entail maintaining another file.  I am strictly wearing my maintainers hat at the moment.

**Comment by sshine** [comment]
@ZapAnton: I've wanted to make a matrix of (*track* x *exercise*) with each cell having a color that indicates whether an exercise is implemented and if its tests are up to date. I thought of using the git timestamp of the test files and the canonical data for the exercise. This is a bit crude, but besides the problem of determining what version a track test has, you also avoid when tracks tests deviate from canonical tests. (I don't know how common it is.)

**Comment by rpottsoh** [comment]
> If we can make the version appear in such a way that this doesn't cause any mental overhead for learners, I'd prefer that.

I am all for that.  So far not a single learner has inquired about the `CanonicalVersion` constant that exists near the top of every test suite of the Delphi track.  Granted, the Delphi track sees nowhere the amount of traffic of the popular tracks.

At the time I chose a constant over a comment to be more flexible going forward; in case a "mechanism" were to be developed between the test suite and the students solutions to help them realize that the test suite had been updated.  But first and foremost, it is a tool for me the maintainer, to make sure that the test suite was current with canonical data since it is easy for canonical data and test suites to become out of sync.

Here is an example of how the constant appears.  I can easily add a comment to inform learners that this constant is nothing they need to be concerned about......

```pascal
unit uCryptoSquareTest;

interface
uses
  DUnitX.TestFramework;

const
  CanonicalVersion = '3.2.0';

type
  [TestFixture]
  TCryptoSquareTest = class(TObject)
  public
    [Test]
//    [Ignore('Comment the "[Ignore]" statement to run the test')]
    procedure empty_plaintext_results_in_an_empty_ciphertext;

    [Test]
    [Ignore]
    procedure Lowercase;
......
```

**Comment by kytrinyx** [comment]
> So far not a single learner has inquired about the CanonicalVersion constant that exists near the top of every test suite of the Delphi track.

Fair enough.

For what it's worth, the website itself will keep track of whether or not the test suite has been updated since the learner started working on the exercise, and will let them update their exercise to use the latest version.

**Comment by sshine** [comment]
@ZapAnton: Have you made this small utility, or have you post-poned or abandoned the idea?

**Comment by ZapAnton** [comment]
@sshine I did, and I have made a [proposal](https://github.com/exercism/rust/pull/742) to add it to the Rust track. Since the maintainer deemed that it is not practical to add the utility to the track, I will most likely extract the utility in my own separate  Github repo

**Comment by sshine** [comment]
There may be a better way to socialize these tools; @bitfield wrote a tool for automating mentor feedback that hasn't escaped the Go track as far as I know. I've written some hackish tool to extract statistics from the mentor queue. It is likely that we will be the sole maintainers of these prototypical tools, but we should at least gather and share knowledge of their existence; if not to spread their use, then at least share ideas.

Perhaps we can start with individual GitHub repos and a Slack channel for tool makers?

**Comment by sshine** [comment]
@ZapAnton: There's a #toolmakers Slack channel now.

**Comment by kotp** [comment]
[Toolmakers slack channel link](https://exercism-team.slack.com/messages/CEDATRP6U/)
